apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+essential to discuss the workings of the three layers within the compilation
+stack, namely `dynamo`, `aot_autograd`, and `inductor`.
+
+The `dynamo` and `aot_autograd` layers primarily focus on capturing the
+computation graph and converting the captured operations into more basic operations.
+This captured graph is then passed to `inductor`, the compiler. `inductor` then applies
+various optimizations passes before generating specialized code.
+
+To gain insight into the different stages of this stack,
+let us compile a simple program in debug mode.
+
+```python
+# Run this file with `TORCH_COMPILE_DEBUG=1`
+
+import torch
+
+def fn(x):
+    return torch.sin(x) + torch.square(x)
+
+torch.compile(fn)(torch.randn(3, 3))
+```
+
+**dynamo**: The primary responsibility of dynamo is to trace the Python program
+and convert it into the FX graph format. The FX graph generated by `dynamo`
+represents PyTorch operations from the public API, such as `torch.sin`.
+Below, you can observe the graph captured by `dynamo`.
+
+```python
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_ : torch.Tensor):
+        l_x_ = L_x_
+
+        # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+        sin = torch.sin(l_x_)
+        square = torch.square(l_x_);  l_x_ = None
+        add = sin + square;  sin = square = None
+        return (add,)
+```
+
+**aot_autograd**: `aot_autograd` retraces all PyTorch operations to
+produce a lower-level FX graph using `aten` functions (from the private API).
+Additionally, `aot_autograd` decomposes composite operations into primitive
+operations. For instance, a composite operation like `torch.square` is traced
+down to `torch.pow(x, 2)`.
+
+Moreover, `aot_autograd` also manages the creation of the backward graph when
+requested. This is useful for transforms like `grad`, `vjp`, etc. Below, you can see the
+graph generated by `aot_autograd` for the above program.
+
+```python
+def forward(self, arg0_1: f32[3, 3]):
+    # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+    sin: f32[3, 3] = torch.ops.aten.sin.default(arg0_1)
+    pow_1: f32[3, 3] = torch.ops.aten.pow.Tensor_Scalar(arg0_1, 2);  arg0_1 = None
+    add: f32[3, 3] = torch.ops.aten.add.Tensor(sin, pow_1);  sin = pow_1 = None
+    return (add,)
+
+```
+
+**inductor**: As discussed above, it is inductor's job to apply optimizations and
+generate specialized code. In this case, it has fused `sin` and `square` to run
+within the same `for`-loop. This allows the generated program to do more compute
+per read/write effectively improving the memory bandwidth utilization.
+
+```c++
+#include "/tmp/torchinductor_kshiteej/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
+extern "C" void kernel(const float* in_ptr0, float* out_ptr0) {
+  for (long i0 = 0L; i0 < 8L; i0 += 8L) {
+    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + i0);
+    auto tmp1 = tmp0.sin();
+    auto tmp2 = tmp0.cos();
```

Comment:
Is this code from a different example? There's no `cos` in your python code.

Also would the triton code be nicer to show here?

Reply:
Ouch, this is from another example. I will try triton code and see which seems better. Thanks!

Reply:
I think C++ code is more readable. Wdyt?

Triton (I think this is nice without the `pointwise` decorator)
```python
@pointwise(size_hints=[16], filename=__file__, meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'mutated_arg_names': [], 'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_add_pow_sin_0', 'configs': [instance_descriptor(divisible_by_16=(0, 1), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=())]})
@triton.jit
def triton_(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 9
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.sin(tmp0)
    tmp2 = tmp0 * tmp0
    tmp3 = tmp1 + tmp2
    tl.store(out_ptr0 + (x0), tmp3, xmask)
```

C++
```c++
#include "/tmp/torchinductor_kshiteej/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const float* in_ptr0, float* out_ptr0) {
  for (long i0 = 0L; i0 < 8L; i0 += 8L) {
    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + i0);
    auto tmp1 = tmp0.sin();
    auto tmp2 = tmp0 * tmp0;
    auto tmp3 = tmp1 + tmp2;
    tmp3.store(out_ptr0 + i0);
  }

#pragma omp simd simdlen(4)
  for (long i0 = 8L; i0 < 9L; i0 += 1L) {
    auto tmp0 = in_ptr0[i0];
    auto tmp1 = std::sin(tmp0);
    auto tmp2 = decltype(tmp0)(tmp0 * tmp0);
    auto tmp3 = tmp1 + tmp2;
    out_ptr0[i0] = tmp3;
  }
}
```

Reply:
I vote for the C++ code but removing the include, and makign the input a multiple of 8, so that we can remove the second loop.

Reply:
I think it's fine to drop the `@pointwise` decorator for the triton code.

Reply:
Have opted for C++ code but with input of size 16.

```c++
extern "C" void kernel(const float* in_ptr0, float* out_ptr0) {
  for (long i0 = 0L; i0 < 16L; i0 += 8L) {
    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + i0);
    auto tmp1 = tmp0.sin();
    auto tmp2 = tmp0 * tmp0;
    auto tmp3 = tmp1 + tmp2;
    tmp3.store(out_ptr0 + i0);
  }
}
```